### PR TITLE
refactor: move security checks to service layer

### LINKED ISSUE #53: Move security checks from the controller layer to the service layer
All calls such as 
```
if !auth.IsSpecificServiceAccount(ctx, auth.Auth) {
	log.Error(ctx, nil, "unauthorized access to cluster info")
	return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("unauthorized access to cluster info"))
}
```
should be handled at the service layer to enforce the security of the application, by avoiding that an endpoint with weak security constraints calls a service method that would return sensitive data.

In the case of listing/showing cluster(s) with sensitive info, the strategy would be to duplicate the existing service methods, check that the caller is allowed to perform the operation and return appropriate level the information, depending on which method was called.

### PR DESCRIPTION
**Pull request by xcoulon**
introduce distinct methods for `Load` and `List`:
- `Load(...)` and `LoadForAuth(...)`
- `List(...)` and `ListForAuth(...)`

Only the `...ForAuth(...)` methods return the sensitive
cluster details.
Those methods are restricted to the `fabric8-auth` SA.

Fixes #53

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

**Comment by xcoulon** [comment]
@alexeykazakov @dipak-pawar this PR only includes changes for the `Load` and `List` endpoints / service methods, but it does not cover the other endpoints which require AuthN/AuthZ (`create cluster`, `link identity to cluster` and `remove cluster/identity link`). I believe that for the sake of uniformity, we should also make those changes, and all subsequent security checks should be done at the service layer level. WDYT? (I'd be happy to do it in a subsequent PR if that's ok for you)

File: cluster/service/cluster_service.go [hunk 1]
```diff
@@ -361,6 +393,44 @@ func (c clusterService) loadClusterByURL(ctx context.Context, clusterURL string)
...
+		c.SAUsername = ""
+		c.SATokenEncrypted = false
+		c.TokenProviderID = ""
+		// need to replace entry in slice since it's not a slice of pointers
+		clusters[i] = c
+	}
+	return clusters, nil
+}
+
+// List lists ALL clusters, including sensitive informatio
```
Comment (alexeykazakov) [inline cluster/service/cluster_service.go]: ... informatio**n**

File: cluster/service/cluster_service.go [hunk 2]
```diff
@@ -85,9 +87,39 @@ func (c clusterService) CreateOrSaveCluster(ctx context.Context, clustr *reposit
...
+	if err != nil {
+		return nil, err
+	}
+	// hide all sensitive info from the cluster record to return
+	result.AuthDefaultScope = ""
+	result.AuthClientID = ""
+	result.AuthClientSecret = ""
+	result.SAToken = ""
+	result.SAUsername = ""
+	result.TokenProviderID = ""
```
Comment (alexeykazakov) [inline cluster/service/cluster_service.go]: c.SATokenEncrypted = false

**Comment by alexeykazakov** [comment]
> I believe that for the sake of uniformity, we should also make those changes, and all subsequent security checks should be done at the service layer level. WDYT? (I'd be happy to do it in a subsequent PR if that's ok for you)

I'm fine with both options. We could keep them as is or unify our security checks in all endpoints. Not the highest priority to unify it right now though ;)

File: cluster/service/cluster_service.go (hunk 2 above)
Comment (xcoulon) [reply cluster/service/cluster_service.go]: fixed in 91e1b5b

File: cluster/service/cluster_service.go (hunk 1 above)
Comment (xcoulon) [reply cluster/service/cluster_service.go]: oops :/ 
fixed in 91e1b5b

**Comment by xcoulon** [comment]
> > I believe that for the sake of uniformity, we should also make those changes, and all subsequent security checks should be done at the service layer level. WDYT? (I'd be happy to do it in a subsequent PR if that's ok for you)
> 
> I'm fine with both options. We could keep them as is or unify our security checks in all endpoints. Not the highest priority to unify it right now though ;)

ok, let me open a new issue for those other methods, so this PR can remain as-is ;)

**Comment by xcoulon** [comment]
thanks @alexeykazakov

**Pull request merged**